File: 2022_11_03_leetcode_in_fsharp/LeetCode/LeetCode/P0066_PlusOne.fs
// // https://leetcode.com/problems/plus-one/

module LeetCode.P0066_PlusOne

let digits = [ 4; 3; 2; 2 ]

let plusOne digits =
    Seq.append [ 0 ] digits
    |> Seq.rev
    |> Seq.scan
        (fun (_, carry) digit ->
            if carry = 0 then digit, 0
            else if digit = 9 then 0, 1
            else digit + 1, 0)
        (0, 1)
    |> Seq.skip 1
    |> Seq.map fst
    |> Seq.rev
    |> Seq.skipWhile (fun d -> d = 0)
    |> Seq.toArray

let _ = plusOne [ 4; 3; 2; 2 ] // -> [|4; 3; 2; 3|]
let _ = plusOne [ 1; 2; 3 ] // -> [|1; 2; 4|]
let _ = plusOne [ 9 ] // -> [|1; 0|]

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0066_PlusOne.cs
// https://leetcode.com/problems/plus-one/

namespace LeetCode;

using static System.Linq.Enumerable;

class P0066_PlusOne
{
    static int[] PlusOne(int[] digits)
    {
        return new[] { 0 }
            .Concat(digits)
            .Reverse()
            .Scan(
                new { NextDigit = 0, Carry = 1 },
                (s, digit) => (s.Carry, digit) switch
                {
                    (0, _) => new { NextDigit = digit, Carry = 0 },
                    (_, 9) => new { NextDigit = 0, Carry = 1 },
                    _ => new { NextDigit = digit + 1, Carry = 0 }
                }
            )
            .Skip(1)
            .Select(s => s.NextDigit)
            .Reverse()
            .SkipWhile(d => d == 0)
            .ToArray();
    }

    public static void Run()
    {
        var inputs = new[]
        {
            new [] { 4, 3, 2, 2},
            new [] { 1, 2, 3},
            new [] { 9},
        };

        foreach (var input in inputs)
        {
            var output = PlusOne(input);
            Console.WriteLine($" {input.JoinToString()} -> {output.JoinToString()}");
        }
    }
}
